src/analysis/ring_seminorm.lean

```diff
@@ -0,0 +1,188 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: María Inés de Frutos-Fernández
+-/
+import analysis.seminorm
+
+/-!
+# Seminorms and norms on rings
+
+This file defines seminorms and norms on rings. These definitions are useful when one needs to
+consider multiple (semi)norms on a given ring.
+
+## Main declarations
+
+For a ring `R`:
+* `ring_seminorm`: A seminorm on a ring `R` is a function `f : R → ℝ` that preserves zero, takes
+  nonnegative values, is subadditive and submultiplicative and such that `f (-x) = f x` for all
+  `x ∈ R`.
+* `ring_norm`: A seminorm `f` is a norm if `f x = 0` if and only if `x = 0`.
+
+## References
+
+* [S. Bosch, U. Güntzer, R. Remmert, *Non-Archimedean Analysis*][bosch-guntzer-remmert]
+
+## Tags
+ring_seminorm, ring_norm
+-/
+
+noncomputable theory
+
+set_option old_structure_cmd true
+
+open_locale nnreal
+
+variables {R S : Type*} (x y : R) (r : ℝ)
+
+/-- A seminorm on a ring `R` is a function `f : R → ℝ` that preserves zero, takes nonnegative
+  values, is subadditive and submultiplicative and such that `f (-x) = f x` for all `x ∈ R`. -/
+structure ring_seminorm (R : Type*) [ring R]
+  extends add_group_seminorm R :=
+(mul_le' : ∀ x y : R, to_fun (x * y) ≤ to_fun x * to_fun y)
+
+attribute [nolint doc_blame] ring_seminorm.to_add_group_seminorm
+
+namespace ring_seminorm
+
+variables [ring R]
+
+/-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`. -/
+instance : has_coe_to_fun (ring_seminorm R) (λ _, R → ℝ) := ⟨λ p, p.to_fun⟩
+
+instance zero_hom_class : zero_hom_class (ring_seminorm R) R ℝ :=
+ { coe := λ f, f.to_fun,
+   coe_injective' := λ f g h, by cases f; cases g; congr',
+   map_zero := λ f, f.map_zero' }
+
+@[ext] lemma ext {p q : ring_seminorm R} (h : ∀ x, p x = q x) : p = q :=
+fun_like.ext p q h
+
+instance : has_zero (ring_seminorm R) :=
+⟨{ mul_le' :=  λ _ _, eq.ge (zero_mul _),
+  ..add_group_seminorm.has_zero.zero }⟩
+
+instance : inhabited (ring_seminorm R) := ⟨0⟩
+
+variables (p : ring_seminorm R)
+
+protected lemma nonneg : 0 ≤ p x := p.nonneg' _
+@[simp] protected lemma map_zero : p 0 = 0 := p.map_zero'
+protected lemma add_le : p (x + y) ≤ p x + p y := p.add_le' _ _
+@[simp] protected lemma neg : p (- x) = p x := p.neg' _
+protected lemma mul_le : p (x * y) ≤ p x * p y := p.mul_le' _ _
+
+lemma pow_le : ∀ {n : ℕ}, 0 < n → p (x ^ n) ≤ (p x) ^ n
+| 1 h := by simp only [pow_one]
+| (n + 2) h :=
+by simpa only [pow_succ _ (n + 1)] using le_trans (p.mul_le x _)
+  (mul_le_mul_of_nonneg_left (pow_le n.succ_pos) (p.nonneg _))
+
+/-- A function `f : R → ℝ` from a ring `R` satisfies `is_norm_le_one_class` if `f 1 ≤ 1`. -/
+def is_norm_le_one_class {R : Type*} [semiring R] (f : R → ℝ) : Prop := f 1 ≤ 1
+
+/-- A function `f : R → ℝ` from a ring `R` satisfies `is_norm_one_class` if `f 1 = 1`. -/
+def is_norm_one_class {R : Type*} [semiring R] (f : R → ℝ) : Prop := f 1 = 1
```

Comment:
I don't think we should have either of these:
* The use of `class` in the name is misleading, these are not `class`es
* The use of `norm` in the name is misleading, these do not mention anything about the norm
* `is_norm_one_class f` is  longer to spell than `f 1 = 1`
* `is_norm_one_class` is longer to spell (as part of a lemma name) than `map_one_eq`

Reply:
Ok, I don't have a strong preference for this, I've removed them.